Nil node on tree

```DEBUG[07-26|09:01:41] orderbook restored                       orderBook.Item="&{NextOrderID:6 MaxPricePoint:0 Name:0x9a8531c62d02af08cf237eb8aecae9dbcb69b6fd::0x9a8531c62d02af08cf237eb8aecae9dbcb69b6fd}"
INFO [07-26|09:01:41] Process order pending                    orderPending="&{Quantity:+1000000000000000000 Price:+200000000000000000 ExchangeAddress:[48 48 48 48 48 48 48 48 48 48 48 48 48 48 48 48 48 48 48 48] UserAddress:[97 53 49 102 56 52 100 57 97 100 99 54 55 98 57 50 49 98 100 102] BaseToken:[55 101 98 56 97 101 99 97 101 57 100 98 99 98 54 57 98 54 102 100] QuoteToken:[55 101 98 56 97 101 99 97 101 57 100 98 99 98 54 57 98 54 102 100] Status:New Side:SELL Type:LO Hash:[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 55] Signature:0xc00bfee8c0 FilledAmount:+0 Nonce:+7 MakeFee:+4000000000000000 TakeFee:+4000000000000000 PairName:0x9a8531c62d02af08cf237eb8aecae9dbcb69b6fd::0x9a8531c62d02af08cf237eb8aecae9dbcb69b6fd CreatedAt:1564131696 UpdatedAt:1564131696 OrderID:0 NextOrder:[] PrevOrder:[] OrderList:[] Key:}"
DEBUG[07-26|09:01:41] M1 - item info                           obOld="&{NextOrderID:6 MaxPricePoint:0 Name:0x9a8531c62d02af08cf237eb8aecae9dbcb69b6fd::0x9a8531c62d02af08cf237eb8aecae9dbcb69b6fd}" askOld="&{Volume:+0 NumOrders:0 PriceTreeKey:[] PriceTreeSize:0}" bidOld="&{Volume:+4000000000000000000 NumOrders:4 PriceTreeKey:[79 76 40 218 224 180 125 197 205 56 148 110 240 228 38 102 179 120 3 60 163 183 110 210 155 220 157 23 77 238 105 182 160 130] PriceTreeSize:4}"
DEBUG[07-26|09:01:41] Get order from key                       storedKey="[40 218 224 180 125 197 205 56 148 110 240 228 38 102 179 120 3 60 163 183 109 210 155 220 154 80 194 253 174 162 160 134]"
DEBUG[07-26|09:01:41] Debug DB delete from dry-run cache       cacheKey=28dae0b47dc5cd38946ef0e42666b378033ca3b76dd29bdc9a50c2fdaea2a086
DEBUG[07-26|09:01:41] Get order from key                       storedKey="[40 218 224 180 125 197 205 56 148 110 240 228 38 102 179 120 3 60 163 183 109 210 155 220 154 80 194 253 174 162 160 130]"
DEBUG[07-26|09:01:41] Get order from key                       storedKey="[40 218 224 180 125 197 205 56 148 110 240 228 38 102 179 120 3 60 163 183 109 210 155 220 154 80 194 253 174 162 160 130]"
DEBUG[07-26|09:01:41] Save orderlist                           key=4f4c28dae0b47dc5cd38946ef0e42666b378033ca3b76ed29bdc9fddd8df24caa082 value="[248 77 160 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 160 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 128 128 136 5 141 21 225 118 40 0 0]"
DEBUG[07-26|09:01:41] Debug put to dry-run cache               cacheKey=4f4c28dae0b47dc5cd38946ef0e42666b378033ca3b76ed29bdc9fddd8df24caa082                   val="&{Keys:0xc00be22550 Value:[248 77 160 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 160 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 128 128 136 5 141 21 225 118 40 0 0] Color:true}"
DEBUG[07-26|09:01:41] Debug get from dry-run cache             cacheKey=4f4c28dae0b47dc5cd38946ef0e42666b378033ca3b76ed29bdc9fddd8df24caa082                   val="&{Keys:0xc00be22550 Value:[248 77 160 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 160 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 128 128 136 5 141 21 225 118 40 0 0] Color:true}"
DEBUG[07-26|09:01:41] Debug put to dry-run cache               cacheKey=4f4c28dae0b47dc5cd38946ef0e42666b378033ca3b76ed29bdc9fddd8df24caa082                   val="&{Keys:0xc00be22550 Value:[248 77 160 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 160 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 128 128 136 5 141 21 225 118 40 0 0] Color:false}"
DEBUG[07-26|09:01:41] Debug put to dry-run cache               cacheKey=4f4c28dae0b47dc5cd38946ef0e42666b378033ca3b76ed29bdc9bb408760c2ca082                   val="&{Keys:0xc00be22780 Value:[248 85 160 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 160 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 136 13 224 182 179 167 100 0 0 136 1 99 69 120 93 138 0 0] Color:false}"
DEBUG[07-26|09:01:41] Debug DB delete from dry-run cache       cacheKey=4f4c28dae0b47dc5cd38946ef0e42666b378033ca3b76ed29bdc9d174dee69b6a082
DEBUG[07-26|09:01:41] delete node,  key: %x, parentKey :%x
    LOG15_ERROR=
DEBUG[07-26|09:01:41] Debug DB delete from dry-run cache       cacheKey=4f4c28dae0b47dc5cd38946ef0e42666b378033ca3b76ed29bdc9fddd8df24caa082
DEBUG[07-26|09:01:41] Debug get from dry-run cache             cacheKey=4f4c28dae0b47dc5cd38946ef0e42666b378033ca3b76ed29bdc9d174dee69b6a082                   val=nil
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0xb5e076]

goroutine 62 [running]:
github.com/ethereum/go-ethereum/tomox.(*Tree).Save(...)
	/root/gocode/tomo/src/github.com/ethereum/go-ethereum/build/_workspace/src/github.com/ethereum/go-ethereum/tomox/redblacktree.go:508
github.com/ethereum/go-ethereum/tomox.(*Tree).replaceNode(0xc00bf83280, 0xc00bf91c60, 0xc00bf91ca0, 0x1)
	/root/gocode/tomo/src/github.com/ethereum/go-ethereum/build/_workspace/src/github.com/ethereum/go-ethereum/tomox/redblacktree.go:415 +0x366
github.com/ethereum/go-ethereum/tomox.(*Tree).Remove(0xc00bf83280, 0xc00bd17800, 0x22, 0x30, 0x1)
	/root/gocode/tomo/src/github.com/ethereum/go-ethereum/build/_workspace/src/github.com/ethereum/go-ethereum/tomox/redblacktree.go:186 +0x263
github.com/ethereum/go-ethereum/tomox.(*OrderTree).RemovePrice(0xc00be22280, 0xc00bf91740, 0x1, 0x0, 0x0)
	/root/gocode/tomo/src/github.com/ethereum/go-ethereum/build/_workspace/src/github.com/ethereum/go-ethereum/tomox/ordertree.go:188 +0x110
github.com/ethereum/go-ethereum/tomox.(*OrderTree).RemoveOrderFromOrderList(0xc00be22280, 0xc00bf918c0, 0xc00bd2b1a0, 0x1, 0x1, 0xc00bf918c0)
	/root/gocode/tomo/src/github.com/ethereum/go-ethereum/build/_workspace/src/github.com/ethereum/go-ethereum/tomox/ordertree.go:324 +0x140
github.com/ethereum/go-ethereum/tomox.(*OrderBook).processOrderList(0xc00be22230, 0x11ec819, 0x3, 0xc00bd2b1a0, 0xc00c09ac20, 0xc0000358c0, 0xc0083f0101, 0x0, 0x0, 0x0, ...)
	/root/gocode/tomo/src/github.com/ethereum/go-ethereum/build/_workspace/src/github.com/ethereum/go-ethereum/tomox/orderbook.go:361 +0xc38
github.com/ethereum/go-ethereum/tomox.(*OrderBook).processLimitOrder(0xc00be22230, 0xc0000358c0, 0x101, 0x155, 0x0, 0xc00bb62000, 0xc00c6a2dd0, 0xb4477c, 0x1166300)
	/root/gocode/tomo/src/github.com/ethereum/go-ethereum/build/_workspace/src/github.com/ethereum/go-ethereum/tomox/orderbook.go:279 +0x290
github.com/ethereum/go-ethereum/tomox.(*OrderBook).ProcessOrder(0xc00be22230, 0xc0000358c0, 0xc00b370101, 0x155, 0x155, 0x0, 0x0, 0x30, 0x149ede0)
	/root/gocode/tomo/src/github.com/ethereum/go-ethereum/build/_workspace/src/github.com/ethereum/go-ethereum/tomox/orderbook.go:318 +0xb1
github.com/ethereum/go-ethereum/tomox.(*TomoX).ProcessOrderPending(0xc0000e7680, 0xc00bf83200)
	/root/gocode/tomo/src/github.com/ethereum/go-ethereum/build/_workspace/src/github.com/ethereum/go-ethereum/tomox/tomox.go:839 +0x913
github.com/ethereum/go-ethereum/miner.(*worker).commitNewWork(0xc0002423c0)
	/root/gocode/tomo/src/github.com/ethereum/go-ethereum/build/_workspace/src/github.com/ethereum/go-ethereum/miner/worker.go:612 +0xb24
github.com/ethereum/go-ethereum/miner.(*worker).update(0xc0002423c0)
	/root/gocode/tomo/src/github.com/ethereum/go-ethereum/build/_workspace/src/github.com/ethereum/go-ethereum/miner/worker.go:285 +0x762
created by github.com/ethereum/go-ethereum/miner.newWorker
	/root/gocode/tomo/src/github.com/ethereum/go-ethereum/build/_workspace/src/github.com/ethereum/go-ethereum/miner/worker.go:168 +0x405
^C```

--- a/miner/worker.go
+++ b/miner/worker.go
@@ -609,6 +609,7 @@ func (self *worker) commitNewWork() {
 					}
 				}
 
+				log.Debug("Start processing order pending")
 				txMatches := tomoX.ProcessOrderPending()
 				if len(txMatches) > 0 {
 					log.Debug("transaction matches found", "txMatches", txMatches)

--- a/tomox/orderbook.go
+++ b/tomox/orderbook.go
@@ -253,21 +253,26 @@ func (orderBook *OrderBook) processLimitOrder(order *OrderItem, verbose bool, dr
 
 	if side == Bid {
 		minPrice := orderBook.Asks.MinPrice(dryrun)
+		log.Debug("Min price in asks tree", "price", minPrice.String(), "dryrun", dryrun)
 		for quantityToTrade.Cmp(zero) > 0 && orderBook.Asks.NotEmpty() && price.Cmp(minPrice) >= 0 {
 			bestPriceAsks := orderBook.Asks.MinPriceList(dryrun)
+			log.Debug("Orderlist at min price", "orderlist", bestPriceAsks.Item, "dryrun", dryrun)
 			quantityToTrade, newTrades, orderInBook, err = orderBook.processOrderList(Ask, bestPriceAsks, quantityToTrade, order, verbose, dryrun)
 			if err != nil {
 				return nil, nil, err
 			}
 			trades = append(trades, newTrades...)
+			log.Debug("New trade found", "newTrades", newTrades, "orderInBook", orderInBook, "quantityToTrade", quantityToTrade, "dryrun", dryrun)
 			minPrice = orderBook.Asks.MinPrice(dryrun)
+			log.Debug("New min price in asks tree", "price", minPrice.String(), "dryrun", dryrun)
 		}
 
 		if quantityToTrade.Cmp(zero) > 0 {
 			order.OrderID = orderBook.Item.NextOrderID
 			order.Quantity = quantityToTrade
+			log.Debug("After matching, order (unmatched part) is now added to bids tree", "order", order, "dryrun", dryrun)
 			if err := orderBook.Bids.InsertOrder(order, dryrun); err != nil {
-				log.Error("Failed to insert order to bidTree", "pairName", order.PairName, "orderItem", order, "err", err)
+				log.Error("Failed to insert order to bidTree", "pairName", order.PairName, "orderItem", order, "err", err, "dryrun", dryrun)
 				return nil, nil, err
 			}
 			orderInBook = order
@@ -275,21 +280,26 @@ func (orderBook *OrderBook) processLimitOrder(order *OrderItem, verbose bool, dr
 
 	} else {
 		maxPrice := orderBook.Bids.MaxPrice(dryrun)
+		log.Debug("Max price in bids tree", "price", maxPrice.String(), "dryrun", dryrun)
 		for quantityToTrade.Cmp(zero) > 0 && orderBook.Bids.NotEmpty() && price.Cmp(maxPrice) <= 0 {
 			bestPriceBids := orderBook.Bids.MaxPriceList(dryrun)
+			log.Debug("Orderlist at max price", "orderlist", bestPriceBids.Item, "dryrun", dryrun)
 			quantityToTrade, newTrades, orderInBook, err = orderBook.processOrderList(Bid, bestPriceBids, quantityToTrade, order, verbose, dryrun)
 			if err != nil {
 				return nil, nil, err
 			}
 			trades = append(trades, newTrades...)
+			log.Debug("New trade found", "newTrades", newTrades, "orderInBook", orderInBook, "quantityToTrade", quantityToTrade, "dryrun", dryrun)
 			maxPrice = orderBook.Bids.MaxPrice(dryrun)
+			log.Debug("New max price in bids tree", "price", maxPrice.String(), "dryrun", dryrun)
 		}
 
 		if quantityToTrade.Cmp(zero) > 0 {
 			order.OrderID = orderBook.Item.NextOrderID
 			order.Quantity = quantityToTrade
+			log.Debug("After matching, order (unmatched part) is now back to asks tree", "order", order, "dryrun", dryrun)
 			if err := orderBook.Asks.InsertOrder(order, dryrun); err != nil {
-				log.Error("Failed to insert order to askTree", "pairName", order.PairName, "orderItem", order, "err", err)
+				log.Error("Failed to insert order to askTree", "pairName", order.PairName, "orderItem", order, "err", err, "dryrun", dryrun)
 				return nil, nil, err
 			}
 			orderInBook = order
@@ -311,11 +321,13 @@ func (orderBook *OrderBook) ProcessOrder(order *OrderItem, verbose bool, dryrun 
 	orderBook.Item.NextOrderID++
 
 	if orderType == Market {
+		log.Debug("Process market order", "order", order)
 		trades, orderInBook, err = orderBook.processMarketOrder(order, verbose, dryrun)
 		if err != nil {
 			return nil, nil, err
 		}
 	} else {
+		log.Debug("Process limit order", "order", order)
 		trades, orderInBook, err = orderBook.processLimitOrder(order, verbose, dryrun)
 		if err != nil {
 			return nil, nil, err
@@ -332,6 +344,7 @@ func (orderBook *OrderBook) ProcessOrder(order *OrderItem, verbose bool, dryrun 
 
 // processOrderList : process the order list
 func (orderBook *OrderBook) processOrderList(side string, orderList *OrderList, quantityStillToTrade *big.Int, order *OrderItem, verbose bool, dryrun bool) (*big.Int, []map[string]string, *OrderItem, error) {
+	log.Debug("Process matching between order and orderlist", "dryrun", dryrun)
 	quantityToTrade := CloneBigInt(quantityStillToTrade)
 	var (
 		trades []map[string]string
@@ -345,6 +358,7 @@ func (orderBook *OrderBook) processOrderList(side string, orderList *OrderList, 
 		if headOrder == nil {
 			return nil, nil, nil, fmt.Errorf("headOrder is null")
 		}
+		log.Debug("Get head order in the orderlist", "headOrder", headOrder.Item, "dryrun", dryrun)
 
 		tradedPrice := CloneBigInt(headOrder.Item.Price)
 
@@ -360,6 +374,7 @@ func (orderBook *OrderBook) processOrderList(side string, orderList *OrderList, 
 			if err := headOrder.UpdateQuantity(orderList, newBookQuantity, headOrder.Item.UpdatedAt, dryrun); err != nil {
 				return nil, nil, nil, err
 			}
+			log.Debug("Update quantity for head order", "headOrder", headOrder.Item, "dryrun", dryrun)
 			quantityToTrade = Zero()
 			orderInBook = headOrder.Item
 		} else if IsEqual(quantityToTrade, headOrder.Item.Quantity) {
@@ -368,10 +383,12 @@ func (orderBook *OrderBook) processOrderList(side string, orderList *OrderList, 
 				if err := orderBook.Bids.RemoveOrderFromOrderList(headOrder, orderList, dryrun); err != nil {
 					return nil, nil, nil, err
 				}
+				log.Debug("Removed headOrder from bids orderlist", "headOrder", headOrder.Item, "orderlist", orderList.Item, "side", side, "dryrun", dryrun)
 			} else {
 				if err := orderBook.Asks.RemoveOrderFromOrderList(headOrder, orderList, dryrun); err != nil {
 					return nil, nil, nil, err
 				}
+				log.Debug("Removed headOrder from asks orderlist", "headOrder", headOrder.Item, "orderlist", orderList.Item, "side", side, "dryrun", dryrun)
 			}
 			quantityToTrade = Zero()
 
@@ -381,10 +398,12 @@ func (orderBook *OrderBook) processOrderList(side string, orderList *OrderList, 
 				if err := orderBook.Bids.RemoveOrderFromOrderList(headOrder, orderList, dryrun); err != nil {
 					return nil, nil, nil, err
 				}
+				log.Debug("Removed headOrder from bids orderlist", "headOrder", headOrder.Item, "orderlist", orderList.Item, "side", side, "dryrun", dryrun)
 			} else {
 				if err := orderBook.Asks.RemoveOrderFromOrderList(headOrder, orderList, dryrun); err != nil {
 					return nil, nil, nil, err
 				}
+				log.Debug("Removed headOrder from asks orderlist", "headOrder", headOrder.Item, "orderlist", orderList.Item, "side", side, "dryrun", dryrun)
 			}
 			quantityToTrade = Sub(quantityToTrade, tradedQuantity)
 		}

--- a/tomox/orderbook.go
+++ b/tomox/orderbook.go
@@ -253,26 +253,26 @@ func (orderBook *OrderBook) processLimitOrder(order *OrderItem, verbose bool, dr
 
 	if side == Bid {
 		minPrice := orderBook.Asks.MinPrice(dryrun)
-		log.Debug("Min price in asks tree", "price", minPrice.String(), "dryrun", dryrun)
+		log.Debug("Min price in asks tree", "price", minPrice.String())
 		for quantityToTrade.Cmp(zero) > 0 && orderBook.Asks.NotEmpty() && price.Cmp(minPrice) >= 0 {
 			bestPriceAsks := orderBook.Asks.MinPriceList(dryrun)
-			log.Debug("Orderlist at min price", "orderlist", bestPriceAsks.Item, "dryrun", dryrun)
+			log.Debug("Orderlist at min price", "orderlist", bestPriceAsks.Item)
 			quantityToTrade, newTrades, orderInBook, err = orderBook.processOrderList(Ask, bestPriceAsks, quantityToTrade, order, verbose, dryrun)
 			if err != nil {
 				return nil, nil, err
 			}
 			trades = append(trades, newTrades...)
-			log.Debug("New trade found", "newTrades", newTrades, "orderInBook", orderInBook, "quantityToTrade", quantityToTrade, "dryrun", dryrun)
+			log.Debug("New trade found", "newTrades", newTrades, "orderInBook", orderInBook, "quantityToTrade", quantityToTrade)
 			minPrice = orderBook.Asks.MinPrice(dryrun)
-			log.Debug("New min price in asks tree", "price", minPrice.String(), "dryrun", dryrun)
+			log.Debug("New min price in asks tree", "price", minPrice.String())
 		}
 
 		if quantityToTrade.Cmp(zero) > 0 {
 			order.OrderID = orderBook.Item.NextOrderID
 			order.Quantity = quantityToTrade
-			log.Debug("After matching, order (unmatched part) is now added to bids tree", "order", order, "dryrun", dryrun)
+			log.Debug("After matching, order (unmatched part) is now added to bids tree", "order", order)
 			if err := orderBook.Bids.InsertOrder(order, dryrun); err != nil {
-				log.Error("Failed to insert order to bidTree", "pairName", order.PairName, "orderItem", order, "err", err, "dryrun", dryrun)
+				log.Error("Failed to insert order to bidTree", "pairName", order.PairName, "orderItem", order, "err", err)
 				return nil, nil, err
 			}
 			orderInBook = order
@@ -280,26 +280,26 @@ func (orderBook *OrderBook) processLimitOrder(order *OrderItem, verbose bool, dr
 
 	} else {
 		maxPrice := orderBook.Bids.MaxPrice(dryrun)
-		log.Debug("Max price in bids tree", "price", maxPrice.String(), "dryrun", dryrun)
+		log.Debug("Max price in bids tree", "price", maxPrice.String())
 		for quantityToTrade.Cmp(zero) > 0 && orderBook.Bids.NotEmpty() && price.Cmp(maxPrice) <= 0 {
 			bestPriceBids := orderBook.Bids.MaxPriceList(dryrun)
-			log.Debug("Orderlist at max price", "orderlist", bestPriceBids.Item, "dryrun", dryrun)
+			log.Debug("Orderlist at max price", "orderlist", bestPriceBids.Item)
 			quantityToTrade, newTrades, orderInBook, err = orderBook.processOrderList(Bid, bestPriceBids, quantityToTrade, order, verbose, dryrun)
 			if err != nil {
 				return nil, nil, err
 			}
 			trades = append(trades, newTrades...)
-			log.Debug("New trade found", "newTrades", newTrades, "orderInBook", orderInBook, "quantityToTrade", quantityToTrade, "dryrun", dryrun)
+			log.Debug("New trade found", "newTrades", newTrades, "orderInBook", orderInBook, "quantityToTrade", quantityToTrade)
 			maxPrice = orderBook.Bids.MaxPrice(dryrun)
-			log.Debug("New max price in bids tree", "price", maxPrice.String(), "dryrun", dryrun)
+			log.Debug("New max price in bids tree", "price", maxPrice.String())
 		}
 
 		if quantityToTrade.Cmp(zero) > 0 {
 			order.OrderID = orderBook.Item.NextOrderID
 			order.Quantity = quantityToTrade
-			log.Debug("After matching, order (unmatched part) is now back to asks tree", "order", order, "dryrun", dryrun)
+			log.Debug("After matching, order (unmatched part) is now back to asks tree", "order", order)
 			if err := orderBook.Asks.InsertOrder(order, dryrun); err != nil {
-				log.Error("Failed to insert order to askTree", "pairName", order.PairName, "orderItem", order, "err", err, "dryrun", dryrun)
+				log.Error("Failed to insert order to askTree", "pairName", order.PairName, "orderItem", order, "err", err)
 				return nil, nil, err
 			}
 			orderInBook = order
@@ -344,7 +344,7 @@ func (orderBook *OrderBook) ProcessOrder(order *OrderItem, verbose bool, dryrun 
 
 // processOrderList : process the order list
 func (orderBook *OrderBook) processOrderList(side string, orderList *OrderList, quantityStillToTrade *big.Int, order *OrderItem, verbose bool, dryrun bool) (*big.Int, []map[string]string, *OrderItem, error) {
-	log.Debug("Process matching between order and orderlist", "dryrun", dryrun)
+	log.Debug("Process matching between order and orderlist")
 	quantityToTrade := CloneBigInt(quantityStillToTrade)
 	var (
 		trades []map[string]string
@@ -358,7 +358,7 @@ func (orderBook *OrderBook) processOrderList(side string, orderList *OrderList, 
 		if headOrder == nil {
 			return nil, nil, nil, fmt.Errorf("headOrder is null")
 		}
-		log.Debug("Get head order in the orderlist", "headOrder", headOrder.Item, "dryrun", dryrun)
+		log.Debug("Get head order in the orderlist", "headOrder", headOrder.Item)
 
 		tradedPrice := CloneBigInt(headOrder.Item.Price)
 
@@ -374,7 +374,7 @@ func (orderBook *OrderBook) processOrderList(side string, orderList *OrderList, 
 			if err := headOrder.UpdateQuantity(orderList, newBookQuantity, headOrder.Item.UpdatedAt, dryrun); err != nil {
 				return nil, nil, nil, err
 			}
-			log.Debug("Update quantity for head order", "headOrder", headOrder.Item, "dryrun", dryrun)
+			log.Debug("Update quantity for head order", "headOrder", headOrder.Item)
 			quantityToTrade = Zero()
 			orderInBook = headOrder.Item
 		} else if IsEqual(quantityToTrade, headOrder.Item.Quantity) {
@@ -383,12 +383,12 @@ func (orderBook *OrderBook) processOrderList(side string, orderList *OrderList, 
 				if err := orderBook.Bids.RemoveOrderFromOrderList(headOrder, orderList, dryrun); err != nil {
 					return nil, nil, nil, err
 				}
-				log.Debug("Removed headOrder from bids orderlist", "headOrder", headOrder.Item, "orderlist", orderList.Item, "side", side, "dryrun", dryrun)
+				log.Debug("Removed headOrder from bids orderlist", "headOrder", headOrder.Item, "orderlist", orderList.Item, "side", side)
 			} else {
 				if err := orderBook.Asks.RemoveOrderFromOrderList(headOrder, orderList, dryrun); err != nil {
 					return nil, nil, nil, err
 				}
-				log.Debug("Removed headOrder from asks orderlist", "headOrder", headOrder.Item, "orderlist", orderList.Item, "side", side, "dryrun", dryrun)
+				log.Debug("Removed headOrder from asks orderlist", "headOrder", headOrder.Item, "orderlist", orderList.Item, "side", side)
 			}
 			quantityToTrade = Zero()
 
@@ -398,12 +398,12 @@ func (orderBook *OrderBook) processOrderList(side string, orderList *OrderList, 
 				if err := orderBook.Bids.RemoveOrderFromOrderList(headOrder, orderList, dryrun); err != nil {
 					return nil, nil, nil, err
 				}
-				log.Debug("Removed headOrder from bids orderlist", "headOrder", headOrder.Item, "orderlist", orderList.Item, "side", side, "dryrun", dryrun)
+				log.Debug("Removed headOrder from bids orderlist", "headOrder", headOrder.Item, "orderlist", orderList.Item, "side", side)
 			} else {
 				if err := orderBook.Asks.RemoveOrderFromOrderList(headOrder, orderList, dryrun); err != nil {
 					return nil, nil, nil, err
 				}
-				log.Debug("Removed headOrder from asks orderlist", "headOrder", headOrder.Item, "orderlist", orderList.Item, "side", side, "dryrun", dryrun)
+				log.Debug("Removed headOrder from asks orderlist", "headOrder", headOrder.Item, "orderlist", orderList.Item, "side", side)
 			}
 			quantityToTrade = Sub(quantityToTrade, tradedQuantity)
 		}

--- a/tomox/orderlist.go
+++ b/tomox/orderlist.go
@@ -88,7 +88,7 @@ func NewOrderListWithItem(item *OrderListItem, orderTree *OrderTree) *OrderList 
 func (orderList *OrderList) GetOrder(key []byte, dryrun bool) *Order {
 	// re-use method from orderbook, because orderlist has the same slot as orderbook
 	storedKey := orderList.GetOrderIDFromKey(key)
-	log.Debug("Get order from key", "storedKey", storedKey)
+	log.Debug("Get order from key", "storedKey", hex.EncodeToString(storedKey))
 	return orderList.orderTree.orderBook.GetOrder(storedKey, key, dryrun)
 }
 
@@ -263,9 +263,9 @@ func (orderList *OrderList) RemoveOrder(order *Order, dryrun bool) error {
 	}
 
 	nextOrder := orderList.GetOrder(order.Item.NextOrder, dryrun)
-	log.Debug("Orderlist remove order debug", "nextOrder", nextOrder.Item)
+	log.Debug("Orderlist remove order debug", "nextOrder", nextOrder)
 	prevOrder := orderList.GetOrder(order.Item.PrevOrder, dryrun)
-	log.Debug("Orderlist remove order debug", "prevOrder", prevOrder.Item)
+	log.Debug("Orderlist remove order debug", "prevOrder", prevOrder)
 
 	orderList.Item.Volume = Sub(orderList.Item.Volume, order.Item.Quantity)
 	orderList.Item.Length--

--- a/tomox/orderlist.go
+++ b/tomox/orderlist.go
@@ -263,7 +263,9 @@ func (orderList *OrderList) RemoveOrder(order *Order, dryrun bool) error {
 	}
 
 	nextOrder := orderList.GetOrder(order.Item.NextOrder, dryrun)
+	log.Debug("Orderlist remove order debug", "nextOrder", nextOrder.Item)
 	prevOrder := orderList.GetOrder(order.Item.PrevOrder, dryrun)
+	log.Debug("Orderlist remove order debug", "prevOrder", prevOrder.Item)
 
 	orderList.Item.Volume = Sub(orderList.Item.Volume, order.Item.Quantity)
 	orderList.Item.Length--
@@ -298,7 +300,6 @@ func (orderList *OrderList) RemoveOrder(order *Order, dryrun bool) error {
 		orderList.Item.HeadOrder = EmptyKey()
 		orderList.Item.TailOrder = EmptyKey()
 	}
-
 	return nil
 }
 

--- a/tomox/ordertree.go
+++ b/tomox/ordertree.go
@@ -184,13 +184,16 @@ func (orderTree *OrderTree) Depth() uint64 {
 func (orderTree *OrderTree) RemovePrice(price *big.Int, dryrun bool) error {
 	if orderTree.Depth() > 0 {
 		priceKey := orderTree.getKeyFromPrice(price)
+		log.Debug("Remove price", "price", price.String(), "priceKey", hex.EncodeToString(priceKey), "dryrun", dryrun)
 		orderListKey := GetOrderListCommonKey(priceKey)
+		log.Debug("Remove price", "price", price.String(), "orderListKey", hex.EncodeToString(orderListKey), "dryrun", dryrun)
 		orderTree.PriceTree.Remove(orderListKey, dryrun)
-
+		log.Debug("Removed price from price tree", "price", price.String(), "orderListKey", hex.EncodeToString(orderListKey), "dryrun", dryrun)
 		// should use batch to optimize the performance
 		if err := orderTree.Save(dryrun); err != nil {
 			return err
 		}
+		log.Debug("Remove price - saved ordertree", "dryrun", dryrun)
 	}
 	return nil
 }
@@ -313,18 +316,20 @@ func (orderTree *OrderTree) RemoveOrderFromOrderList(order *Order, orderList *Or
 	if err := orderList.RemoveOrder(order, dryrun); err != nil {
 		return err
 	}
+	log.Debug("Removed order from orderlist", "order", order.Item, "orderlist", orderList.Item)
 
 	// snapshot order list
 	if err := orderList.Save(dryrun); err != nil {
 		return err
 	}
+	log.Debug("Saved orderlist", "orderlist", orderList.Item)
 
 	// no items left than safety remove
 	if orderList.Item.Length == uint64(0) {
 		if err := orderTree.RemovePrice(order.Item.Price, dryrun); err != nil {
 			return err
 		}
-		log.Debug("remove price list", "price", order.Item.Price.String())
+		log.Debug("Removed price list", "price", order.Item.Price.String())
 	}
 
 	// update orderTree
@@ -398,14 +403,18 @@ func (orderTree *OrderTree) MaxPrice(dryrun bool) *big.Int {
 // MinPrice : get the min price
 func (orderTree *OrderTree) MinPrice(dryrun bool) *big.Int {
 	if orderTree.Depth() > 0 {
+		log.Debug("ordertree isn't empty")
 		if bytes, found := orderTree.PriceTree.GetMin(dryrun); found {
 			item, err := orderTree.getOrderListItem(bytes)
 			if err != nil {
+				log.Error("Failed to get orderlist with min price", "err", err)
 				return Zero()
 			}
 			if item != nil {
 				return CloneBigInt(item.Price)
 			}
+		} else {
+			log.Debug("Min not found")
 		}
 	}
 	return Zero()

--- a/tomox/ordertree.go
+++ b/tomox/ordertree.go
@@ -184,16 +184,16 @@ func (orderTree *OrderTree) Depth() uint64 {
 func (orderTree *OrderTree) RemovePrice(price *big.Int, dryrun bool) error {
 	if orderTree.Depth() > 0 {
 		priceKey := orderTree.getKeyFromPrice(price)
-		log.Debug("Remove price", "price", price.String(), "priceKey", hex.EncodeToString(priceKey), "dryrun", dryrun)
+		log.Debug("Remove price", "price", price.String(), "priceKey", hex.EncodeToString(priceKey))
 		orderListKey := GetOrderListCommonKey(priceKey)
-		log.Debug("Remove price", "price", price.String(), "orderListKey", hex.EncodeToString(orderListKey), "dryrun", dryrun)
+		log.Debug("Remove price", "price", price.String(), "orderListKey", hex.EncodeToString(orderListKey))
 		orderTree.PriceTree.Remove(orderListKey, dryrun)
-		log.Debug("Removed price from price tree", "price", price.String(), "orderListKey", hex.EncodeToString(orderListKey), "dryrun", dryrun)
+		log.Debug("Removed price from price tree", "price", price.String(), "orderListKey", hex.EncodeToString(orderListKey))
 		// should use batch to optimize the performance
 		if err := orderTree.Save(dryrun); err != nil {
 			return err
 		}
-		log.Debug("Remove price - saved ordertree", "dryrun", dryrun)
+		log.Debug("Remove price - saved ordertree")
 	}
 	return nil
 }

--- a/tomox/redblacktree.go
+++ b/tomox/redblacktree.go
@@ -10,6 +10,7 @@ import (
 	"fmt"
 
 	"github.com/ethereum/go-ethereum/log"
+	"encoding/hex"
 )
 
 // Tree holds elements of the red-black tree
@@ -39,7 +40,11 @@ func NewWithBytesComparator(db OrderDao) *Tree {
 }
 
 func (tree *Tree) Root(dryrun bool) *Node {
-	root, _ := tree.GetNode(tree.rootKey, dryrun)
+	root, err := tree.GetNode(tree.rootKey, dryrun)
+	if err != nil {
+		log.Error("Can't get tree.Root", "rootKey", hex.EncodeToString(tree.rootKey), "err", err)
+		return nil
+	}
 	return root
 }
 
@@ -150,10 +155,10 @@ func (tree *Tree) Get(key []byte, dryrun bool) (value []byte, found bool) {
 func (tree *Tree) Remove(key []byte, dryrun bool) {
 	var child *Node
 	node, err := tree.GetNode(key, dryrun)
-
 	if err != nil || node == nil {
 		return
 	}
+	log.Debug("Get node", "node", node, "dryrun", dryrun)
 
 	var left, right *Node = nil, nil
 	if !tree.IsEmptyKey(node.LeftKey()) {
@@ -389,6 +394,7 @@ func (tree *Tree) rotateRight(node *Node, dryrun bool) {
 }
 
 func (tree *Tree) replaceNode(old *Node, new *Node, dryrun bool) {
+	log.Debug("Replace node", "old", old, "new", new, "dryrun", dryrun)
 
 	// we do not change any byte of Key so we can copy the reference to save directly to db
 	var newKey []byte
@@ -410,8 +416,9 @@ func (tree *Tree) replaceNode(old *Node, new *Node, dryrun bool) {
 			oldParent.RightKey(newKey)
 		}
 		// we can have case like: remove a node, then add it again
-		tree.Save(oldParent, dryrun)
-		// }
+		if oldParent != nil {
+			tree.Save(oldParent, dryrun)
+		}
 	}
 	if new != nil {
 		// here is the swap, not update key
@@ -503,10 +510,12 @@ func (tree *Tree) insertCase5(node *Node, dryrun bool) {
 }
 
 func (tree *Tree) Save(node *Node, dryrun bool) error {
+	log.Debug("Save node", "node", node, "dryrun", dryrun)
 	return tree.db.Put(node.Key, node.Item, dryrun)
 }
 
 func (tree *Tree) deleteCase1(node *Node, dryrun bool) {
+	log.Debug("delete case 1", "node value", hex.EncodeToString(node.Value()))
 	if tree.IsEmptyKey(node.ParentKey()) {
 		tree.deleteNode(node, dryrun)
 		return
@@ -516,6 +525,7 @@ func (tree *Tree) deleteCase1(node *Node, dryrun bool) {
 }
 
 func (tree *Tree) deleteCase2(node *Node, dryrun bool) {
+	log.Debug("delete case 2", "node value", hex.EncodeToString(node.Value()))
 	parent := node.Parent(tree, dryrun)
 	sibling := node.sibling(tree, dryrun)
 
@@ -535,7 +545,7 @@ func (tree *Tree) deleteCase2(node *Node, dryrun bool) {
 }
 
 func (tree *Tree) deleteCase3(node *Node, dryrun bool) {
-
+	log.Debug("delete case 3", "node value", hex.EncodeToString(node.Value()))
 	parent := node.Parent(tree, dryrun)
 	sibling := node.sibling(tree, dryrun)
 	siblingLeft := sibling.Left(tree, dryrun)
@@ -548,9 +558,7 @@ func (tree *Tree) deleteCase3(node *Node, dryrun bool) {
 		sibling.Item.Color = red
 		tree.Save(sibling, dryrun)
 		tree.deleteCase1(parent, dryrun)
-		log.Debug("delete node,  key: %x, parentKey :%x\n", node.Key, parent.Key)
 		tree.deleteNode(node, dryrun)
-
 	} else {
 		tree.deleteCase4(node, dryrun)
 	}
@@ -558,6 +566,7 @@ func (tree *Tree) deleteCase3(node *Node, dryrun bool) {
 }
 
 func (tree *Tree) deleteCase4(node *Node, dryrun bool) {
+	log.Debug("delete case 4", "node value", hex.EncodeToString(node.Value()))
 	parent := node.Parent(tree, dryrun)
 	sibling := node.sibling(tree, dryrun)
 	siblingLeft := sibling.Left(tree, dryrun)
@@ -577,6 +586,7 @@ func (tree *Tree) deleteCase4(node *Node, dryrun bool) {
 }
 
 func (tree *Tree) deleteCase5(node *Node, dryrun bool) {
+	log.Debug("delete case 5", "node value", hex.EncodeToString(node.Value()))
 	parent := node.Parent(tree, dryrun)
 	sibling := node.sibling(tree, dryrun)
 	siblingLeft := sibling.Left(tree, dryrun)
@@ -612,6 +622,7 @@ func (tree *Tree) deleteCase5(node *Node, dryrun bool) {
 }
 
 func (tree *Tree) deleteCase6(node *Node, dryrun bool) {
+	log.Debug("delete case 6", "node value", hex.EncodeToString(node.Value()))
 	parent := node.Parent(tree, dryrun)
 	sibling := node.sibling(tree, dryrun)
 	siblingLeft := sibling.Left(tree, dryrun)
@@ -645,5 +656,6 @@ func nodeColor(node *Node) bool {
 }
 
 func (tree *Tree) deleteNode(node *Node, dryrun bool) {
+	log.Debug("Delete node", "node value", hex.EncodeToString(node.Value()), "dryrun", dryrun)
 	tree.db.Delete(node.Key, dryrun)
 }

--- a/tomox/redblacktree.go
+++ b/tomox/redblacktree.go
@@ -178,19 +178,22 @@ func (tree *Tree) Remove(key []byte, dryrun bool) {
 		} else {
 			child = right
 		}
+		if child == nil {
+			tree.deleteNode(node, dryrun)
+		} else {
+			if node.Item.Color == black {
+				node.Item.Color = nodeColor(child)
+				tree.Save(node, dryrun)
 
-		if node.Item.Color == black {
-			node.Item.Color = nodeColor(child)
-			tree.Save(node, dryrun)
+				tree.deleteCase1(node, dryrun)
+			}
 
-			tree.deleteCase1(node, dryrun)
-		}
+			tree.replaceNode(node, child, dryrun)
 
-		tree.replaceNode(node, child, dryrun)
-
-		if tree.IsEmptyKey(node.ParentKey()) && child != nil {
-			child.Item.Color = black
-			tree.Save(child, dryrun)
+			if tree.IsEmptyKey(node.ParentKey()) && child != nil {
+				child.Item.Color = black
+				tree.Save(child, dryrun)
+			}
 		}
 	}
 

--- a/tomox/redblacktree.go
+++ b/tomox/redblacktree.go
@@ -158,7 +158,7 @@ func (tree *Tree) Remove(key []byte, dryrun bool) {
 	if err != nil || node == nil {
 		return
 	}
-	log.Debug("Get node", "node", node, "dryrun", dryrun)
+	log.Debug("Get node", "node", node)
 
 	var left, right *Node = nil, nil
 	if !tree.IsEmptyKey(node.LeftKey()) {
@@ -367,6 +367,7 @@ func output(tree *Tree, node *Node, prefix string, isTail bool, str *string, dry
 }
 
 func (tree *Tree) rotateLeft(node *Node, dryrun bool) {
+	log.Debug("Rotate left - before", "nodeRoot", hex.EncodeToString(tree.Root(dryrun).Value()))
 	right := node.Right(tree, dryrun)
 	tree.replaceNode(node, right, dryrun)
 	node.RightKey(right.LeftKey())
@@ -379,9 +380,11 @@ func (tree *Tree) rotateLeft(node *Node, dryrun bool) {
 	node.ParentKey(right.Key)
 	tree.Save(node, dryrun)
 	tree.Save(right, dryrun)
+	log.Debug("Rotate left - after", "nodeRoot", hex.EncodeToString(tree.Root(dryrun).Value()))
 }
 
 func (tree *Tree) rotateRight(node *Node, dryrun bool) {
+	log.Debug("Rotate right - before", "nodeRoot", hex.EncodeToString(tree.Root(dryrun).Value()))
 	left := node.Left(tree, dryrun)
 	tree.replaceNode(node, left, dryrun)
 	node.LeftKey(left.RightKey())
@@ -394,10 +397,11 @@ func (tree *Tree) rotateRight(node *Node, dryrun bool) {
 	node.ParentKey(left.Key)
 	tree.Save(node, dryrun)
 	tree.Save(left, dryrun)
+	log.Debug("Rotate right - after", "nodeRoot", hex.EncodeToString(tree.Root(dryrun).Value()))
 }
 
 func (tree *Tree) replaceNode(old *Node, new *Node, dryrun bool) {
-	log.Debug("Replace node", "old", old, "new", new, "dryrun", dryrun)
+	log.Debug("Replace node", "old", old, "new", new)
 
 	// we do not change any byte of Key so we can copy the reference to save directly to db
 	var newKey []byte
@@ -513,7 +517,7 @@ func (tree *Tree) insertCase5(node *Node, dryrun bool) {
 }
 
 func (tree *Tree) Save(node *Node, dryrun bool) error {
-	log.Debug("Save node", "node", node, "dryrun", dryrun)
+	log.Debug("Save node", "node", node)
 	return tree.db.Put(node.Key, node.Item, dryrun)
 }
 
@@ -659,6 +663,6 @@ func nodeColor(node *Node) bool {
 }
 
 func (tree *Tree) deleteNode(node *Node, dryrun bool) {
-	log.Debug("Delete node", "node value", hex.EncodeToString(node.Value()), "dryrun", dryrun)
+	log.Debug("Delete node", "node value", hex.EncodeToString(node.Value()))
 	tree.db.Delete(node.Key, dryrun)
 }

--- a/tomox/snapshot.go
+++ b/tomox/snapshot.go
@@ -104,7 +104,7 @@ func prepareOrderTreeData(tree *OrderTree) (*OrderTreeSnapshot, error) {
 	snap.OrderTreeItem = serializedTree
 
 	snap.OrderList = make(map[common.Hash]*OrderListSnapshot)
-	if !tree.NotEmpty() {
+	if !tree.NotEmpty() || tree.PriceTree.Size() == 0 {
 		return snap, nil
 	}
 	// foreach each price, snapshot its orderlist

--- a/tomox/tomox.go
+++ b/tomox/tomox.go
@@ -793,6 +793,7 @@ func (tomox *TomoX) ProcessOrderPending() map[common.Hash]TxDataMatch {
 	txMatches := make(map[common.Hash]TxDataMatch)
 	tomox.db.InitDryRunMode()
 
+	log.Debug("Get pending hashes")
 	pendingHashes := tomox.getPendingHashes()
 	if len(pendingHashes) > 0 {
 		for i, orderHash := range pendingHashes {

